Yubico.YubiKey/src/Yubico/YubiKey/Piv/PivSession.ManagementKey.cs

```diff
@@ -1061,5 +1013,68 @@ private bool TryAuthenticateManagementKey(bool mutualAuthentication,
 
             return ManagementKeyAuthenticated;
         }
+
+        private void RefreshManagementKeyAlgorithm() => ManagementKeyAlgorithm = GetManagementKeyAlgorithm();
+
+        private PivAlgorithm GetManagementKeyAlgorithm()
+        {
+            var response = Connection.SendCommand(new GetMetadataCommand(PivSlot.Management));
+            if (response.Status != ResponseStatus.Success)
+            {
+                throw new InvalidOperationException(response.StatusMessage);
+            }
+
+            var metadata = response.GetData();
+            return metadata.Algorithm;
+        }
+
+        // Verify that and that the given algorithm is allowed.
+        // If checkMode is true, also check that the PIN-only mode is None.
+        // This is called by methods that set PIN-only mode or change the mgmt
+        // key.
+        // The algorithm can only be 3DES or AES, and it can only be AES if the
+        // YubiKey is 5.4.2 or later.
+        // It is not allowed to change the mgmt key if it is PIN-only, so those
+        // methods that change, will check the mode as well (they will pass true
+        // as the checkMode arg).
+        // If setting PIN-only, then the mode is not an issue, so don't check
+        // (pass false as the checkMode arg).
+        // If everything is fine, return, otherwise throw an exception.
+        private void CheckManagementKeyAlgorithm(PivAlgorithm algorithm, bool checkMode)
+        {
+            if (checkMode)
+            {
+                var pinOnlyMode = GetPinOnlyMode();
+                if (pinOnlyMode.HasFlag(PivPinOnlyMode.PinProtected) ||
+                    pinOnlyMode.HasFlag(PivPinOnlyMode.PinDerived))
+                {
+                    throw new InvalidOperationException(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            ExceptionMessages.MgmtKeyCannotBeChanged));
+                }
+            }
+
+            bool isValid = IsValid(algorithm);
+            if (!isValid)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        ExceptionMessages.UnsupportedAlgorithm));
+            }
+
+            return;
+
+            bool IsValid(PivAlgorithm pa) =>
+                pa switch
+                {
+                    PivAlgorithm.TripleDes => true, // Default for keys below fw version 5.7
+                    PivAlgorithm.Aes128 => _yubiKeyDevice.HasFeature(YubiKeyFeature.PivAesManagementKey),
+                    PivAlgorithm.Aes192 => _yubiKeyDevice.HasFeature(YubiKeyFeature.PivAesManagementKey),
+                    PivAlgorithm.Aes256 => _yubiKeyDevice.HasFeature(YubiKeyFeature.PivAesManagementKey),
+                    _ => false
+                };
+        }
```

Comment:
Interesting way of defining scoped functions which I have never seen before. It made me stop scrolling the PR ;)

Reply:
Haha. You were shocked! :D There are even more powerful in C# 13 (We are at version 8)